google/resource_sql_database_instance.go

```diff
@@ -128,24 +145,10 @@ func resourceSqlDatabaseInstance() *schema.Resource {
 							Elem: &schema.Resource{
 								Schema: map[string]*schema.Schema{
 									"authorized_networks": &schema.Schema{
-										Type:     schema.TypeList,
+										Type:     schema.TypeSet,
 										Optional: true,
-										Elem: &schema.Resource{
-											Schema: map[string]*schema.Schema{
-												"expiration_time": &schema.Schema{
-													Type:     schema.TypeString,
-													Optional: true,
-												},
-												"name": &schema.Schema{
-													Type:     schema.TypeString,
-													Optional: true,
-												},
-												"value": &schema.Schema{
-													Type:     schema.TypeString,
-													Optional: true,
-												},
-											},
-										},
+										Set:      schema.HashResource(sqlDatabaseAuthorizedNetWorkSchemaElem),
```

Comment:
This is actually the default behavior, so you don't need this line (I'm fine either way if you want to leave it in to be explicit).

I do wonder whether it actually makes more sense to hash on something a bit more like a primary key, like the network value. Even though our update logic will be the same no matter which part of it changes, it's possible that someone could get confused if they see in a plan that the whole thing is changing rather than just the part of it they changed. What do you think?

Reply:
I made it explicit because below (line 1056), we need to pass the hash function for the NewSet method and by making it explicit here, it is easier to see that both uses the same method.

The API does accept:
```
      authorized_networks {
        name = "stitchdata-3"
        value = "52.204.228.32/32"
        expiration_time = "2018-08-30T13:22:53.108Z"
      }
      authorized_networks {
        name = "stitchdata-3"
        value = "52.204.228.32/32"
        expiration_time = "2018-07-21T13:22:53.108Z"
      }
```

and keeps both entries...

Since the primary key used by the API is all three fields, I suggest we stick to the same key. If we don't, we would be more restrictive than the API which could prevent reasonable use cases that we are not necessary aware of.